akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -887,13 +890,24 @@ private[stream] abstract class MaterializerSession(val topLevel: StreamLayout.Mo
   protected def materializeAtomic(atomic: AtomicModule, effectiveAttributes: Attributes, matVal: ju.Map[Module, Any]): Unit
 
   private def resolveMaterialized(matNode: MaterializedValueNode, matVal: ju.Map[Module, Any], spaces: Int): Any = {
-    if (MaterializerSession.Debug) println(" " * spaces + matNode)
-    val ret = matNode match {
-      case Atomic(m)          ⇒ matVal.get(m)
-      case Combine(f, d1, d2) ⇒ f(resolveMaterialized(d1, matVal, spaces + 2), resolveMaterialized(d2, matVal, spaces + 2))
-      case Transform(f, d)    ⇒ f(resolveMaterialized(d, matVal, spaces + 2))
-      case Ignore             ⇒ NotUsed
+
+    import scala.util.control.TailCalls._
+    def loop(mat: MaterializedValueNode, spaces: Int): TailRec[Any] = {
+      if (MaterializerSession.Debug) println(" " * spaces + mat)
+      mat match {
+        case Atomic(m)          ⇒ done(matVal.get(m))
+        case Combine(f, d1, d2) ⇒
+          for {
+            l <- tailcall(loop(d1, spaces + 2))
+            r <- tailcall(loop(d2, spaces + 2))
+          } yield f(l, r)
+        case Transform(f, d)    ⇒ tailcall(loop(d, spaces + 2)).map(f)
+        case Ignore             ⇒ done(NotUsed)
+      }
     }
+
+
+    val ret = loop(matNode, spaces).result
```

Comment:
I was thinking about that too, but couldn't figure out a way. Maybe the fallback is a good idea, but we should flag that somehow so that we can avoid brute-force-detecting it in both the mat-val traversals and just do it in the first.

Maybe @drewhk knows a better way to know the depth?